### ISSUE TITLE
Handle error with flatMap

### ISSUE THREAD

**Issue by onmyway133** [ReactiveX/RxSwift#1162]
Hi, I have these errors and a flow with `flatMap` and `drive`

```swift
enum MyError: Error {
  case firstFlatMapLatest
  case innerFirstFlatMapLatest
  case secondFlatMapLatest
  case thirdFlatMap
}
```

```swift
button.rx.tap
      .flatMapLatest({
        return Observable.error(MyError.firstFlatMapLatest)
          .flatMapLatest({
            return Observable.error(MyError.innerFirstFlatMapLatest)
          })
          .catchErrorJustReturn(())
      })
      .flatMapLatest({
        return Observable.error(MyError.secondFlatMapLatest).catchErrorJustReturn(())
      })
      .flatMap({
        return Observable.error(MyError.thirdFlatMap).catchErrorJustReturn(())
      })
      .catchError({ error in
        print(error)
        return Observable.just(())
      })
      .map({
        return "hello"
      })
      .asDriver(onErrorRecover: { error in
        print(error)
        return Driver.just("hi")
      })
      .drive(onNext: { value in
        print(value)
      })

```

Here I must use `catchErrorJustReturn` for the flow to not be terminated, but what if I don't do `catchErrorJustReturn`?

- Why can error inside `flatMapLatest` stop the whole flow? In this case `catchError` is only called once
- Why do we need to catch error inside `flatMapLatest`?
- I have an inner `flatMapLatest` inside the first `flatMapLatest`, why don't I need to `catchErrorJustReturn` for that? like

```swift
.flatMapLatest({
    return Observable.error(MyError.innerFirstFlatMapLatest).catchErrorJustReturn(())
})
```

- Is there a difference when using `flatMap` and `flatMapLatest` here

This may related to https://github.com/ReactiveX/RxSwift/issues/618, https://github.com/ReactiveX/RxSwift/issues/729, https://github.com/ReactiveX/RxSwift/issues/304

**Comment by kzaher** [comment]
Hi @onmyway133 ,

The behavior you are describing is a consequence of sequence grammar.
https://github.com/ReactiveX/RxSwift/blob/master/Documentation/GettingStarted.md#getting-started

Safer approach would be to use `Driver` in this chain.

```swift
button.rx.tap.asDriver()
      .flatMapLatest
```

because compiler will create compile time warnings then.

**Comment by onmyway133** [comment]
@kzaher Hi, thanks for your reply. I think we only have a method called `asDriver(onErrorJustReturn)`?

And why don't we need to `catch error` in the inner `flatMapLatest` of the first `flatMapLatest`?

**Comment by kzaher** [comment]
> Hi, thanks for your reply. I think we only have a method called asDriver(onErrorJustReturn)?

Because `ControlProperty`/`ControlEvent`/`Variable` can't produce error, they also have `asDriver()` method.

> And why don't we need to catch error in the inner flatMapLatest of the first flatMapLatest?

The behavior you are describing is a consequence of sequence grammar.
https://github.com/ReactiveX/RxSwift/blob/master/Documentation/GettingStarted.md#getting-started

```
return Observable.error(MyError.firstFlatMapLatest)
          .flatMapLatest({
            return Observable.error(MyError.innerFirstFlatMapLatest)
          })
          .catchErrorJustReturn(()) <--- returns a sequence that doesn't fail
```

**Comment by onmyway133** [comment]
@kzaher Hi, sorry if I ask too much 😇 

In my example, I have `.asDriver(onErrorRecover:`, do I need to do it?

**Comment by kzaher** [comment]
Hi @onmyway133 ,

I'm really sorry, this is an issues channel only. I've tried to help you and point you in the right direction, but please read the docs. Everything is documented regarding this, and there is a lot of examples in this repo.

**Comment by onmyway133** [comment]
@kzaher I read it many times but still don't get it 😢 , they are just very simple examples. @icanzilb maybe you can help me understand in layman 's term?

I think `catchError` is enough to catch the error, but with `flatMap` we need to catch the error inside, which I don't really understand. I can just blindly follow and it works, but an explanation would be great 😇

**Comment by sergdort** [comment]
Hi, @onmyway133 [this](https://github.com/ReactiveX/RxSwift/blob/master/RxSwift/Observables/Merge.swift#L422) and [this](https://github.com/ReactiveX/RxSwift/blob/master/RxSwift/Observables/Merge.swift#L475) is why it does not happened. When you catch inner sequence it completes and decrements number of active sequences but since outer sequence is probably UI related e.g button taps or text field values they are only complete on `deinit` that's why sequence is not stopped

**Comment by sergdort** [comment]
From my experience I can highly recommend reading implementation in case you confused with operator behaviour. It really helped me a lot to understand how everything works

**Comment by onmyway133** [comment]
@sergdort Hi, many thanks for answering. I'm checking it out. Will write a detailed explanation about this once I understand it

**Comment by nahung89** [comment]
The answer from @sergdort is right, but it's a little bit hard to understand. Therefore I make sample codes myself to test it. Basically:
Case_A: when you `catchError` outside `flatMap`, you are transforming outer observable(s) to new observable (that raises error) -> the previous observable(s) will be completed!!!
Case_B: when you `catchError` inside `flatMap`, you chain outer stream into another stream (that raise error but covered ) -> the previous streams continue to run

Note that examples below, I don't call `.addDisposableTo(disposeBag)`. You can see `[TAP]` dispose itself.

```
// Case_A
 _ = buttonLogin.rx.tap.asObservable().debug("[TAP]")
            .flatMap {
                return Observable<Int>.error(MyError.fakeError)
            }.debug("[FLATMAP]")
            .catchErrorJustReturn(0).debug("[JUST]")
            .subscribe()    
        // diagrams    
        // [TAP]              --T->
        // [FLATMAP]    --X->
        // [JUST]           --0-|>
```
```
// Case_B
        _ = buttonLogin.rx.tap.asObservable().debug("[TAP]")
            .flatMap {
                return Observable<Int>.error(MyError.fakeError).debug("[ERROR]")
                                .catchErrorJustReturn(0).debug("[JUST]")
            }.debug("[FLATMAP]")
            .subscribe()
        
        // [TAP]              --T--------T----------T------------T--.....-->
        // [ERROR]        --E->     --E->          --E->            --E->
        // [JUST]           --0-|>    --0-|>        --0-|>           --0-|>
        // [FLATMAP]    --0-------0----------0---------- -0--.....-->
```

Another case is covering error by `catchError( return Observable)`. Then new observable will continue to run, while previous observable(s) will be complete.
```
_ = buttonLogin.rx.tap.asObservable().debug("[TAP]")
            .flatMap {
                return Observable<Int>.error(MyError.fakeError)
            }.debug("[FLATMAP]")
            .catchError { (error) in
                return Observable<Int>.interval(1, scheduler: RxSchedulers.shared.main)
            }.debug("[INTERVAL]")
            .subscribe()
        
        // [TAP]              --T->
        // [FLATMAP]    --X->
        // [INTERVAL]   --1---2---3---4---5---6---7---8---9--.....-->
````

**Comment by jondwillis** [comment]
An alternative to is use catchError and return an empty observable. 

```xObservable.catchError { _ in Observable<TheTypeOfXObservable>.empty() }```
